euphonic/util.py

```diff
@@ -267,13 +267,48 @@ def mode_gradients_to_widths(mode_gradients: Quantity, cell_vectors: Quantity
         raise ValueError(
             f'Unexpected shape for mode_gradients {modg.shape}, '
             f'expected (n_qpts, n_modes) or (n_qpts, n_modes, 3)')
-    cell_volume = Crystal.from_cell_vectors(cell_vectors)._cell_volume()
+    cell_volume = (_cell_vectors_to_volume(
+        cell_vectors.magnitude)*cell_vectors.units**3).to('bohr**3').magnitude
     q_spacing = 2/(np.cbrt(len(mode_gradients)*cell_volume))
     mode_widths = q_spacing*modg
     return mode_widths*ureg('hartree').to(
         mode_gradients.units/cell_vectors.units)
 
 
+def _cell_vectors_to_volume(cell_vectors: np.ndarray) -> float:
+    """Convert 3x3 cell vectors to volume"""
+    return np.dot(cell_vectors[0],
+                  np.cross(cell_vectors[1], cell_vectors[2]))
+
+
+def _get_unique_elems_and_idx(
+        all_elems: Sequence[Union[str, int, Sequence[str], Sequence[int]]]
+    ) -> 'OrderedDict[Union[str, int, tuple], np.ndarray]':
+    """
+    Returns an ordered dictionary mapping the unique elements (or
+    sequences of elements), to their indices
+    """
+    # If each element of all_elems is a sequence, convert each row
+    # to a tuple so it can be compared with np.unique
+    if not isinstance(all_elems[0], (str, int)):
+        tmp = np.empty(len(all_elems), dtype=object)
+        tmp[:] = [tuple(row) for row in all_elems]
+        all_elems = tmp
+
+    _, idx, inverse = np.unique(all_elems, return_index=True,
+                                return_inverse=True)
+    # Retain order
+    sorted_idx = np.sort(idx)
+    inverse_idx = inverse[sorted_idx]
+    unique_elems = all_elems[sorted_idx]
+
+    elem_dict = OrderedDict([
+        (unique_elem, np.where(inverse == inverse_elem_idx)[0])
+        for unique_elem, inverse_elem_idx in zip(unique_elems, inverse_idx)
+        ])
+    return elem_dict
+
```

Comment:
Nice idea! This is also called by `Crystal.get_species_idx`, in that case `len(all_elems) = n_atoms` so I was a little worried with larger numbers of atoms without Numpy it would start to become slow, but this seems to outperform the Numpy implementation up until ~1000 atoms, and by that point we'd have other problems so I'll put it in :)